hack/ci-utils/go_generate.sh

```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+
+set -o errexit
+
+
+main() {
+
+    # Variables checked for timeout values, defaults used if unset
+    soft_timeout="${GO_GENERATE_SOFT_TIMEOUT:-20m}"
+    hard_timeout="${GO_GENERATE_HARD_TIMEOUT:-21m}"
```

Comment:
I landed on 20/21 minutes by incrementing the time required to pass the CI pipeline. My thinking here is we want it to take as little time as possible ideally.
I'm not sure what the longest average successful `go generate` time is. We don't want it to kill early if a few more minutes would've been successful.
How does 30 minutes sound?

When it fails I believe the root cause is slow VM IO performance, since `go generate` is so IO intensive.

Reply:
Maybe a broader question, why fail out the whole pipeline on a timeout when it will eventually be successful or fail out? If we need a pipeline run executed quickly and believe it's stuck, would it be better to just kick off a new pipeline run?

Reply:
So the default timeout value for `go test` is 10 minutes. I believe the thinking here is, your unit tests shouldn't take long, and if they do something is probably wrong.
I think the delay of waiting for a `go generate` run that never completes is more problematic than failing after a known "too long" point. I haven't seen `go generate` succeed past 1 hour, so maybe that could be a timeout value.

Reply:
What is the current behaviour should the `go generate` hang or fail? Does it get stuck indefinitely? We can probably settle on 1 hour here to be on the safe side.

Reply:
@jhoreman It "hangs", or continues to execute very slowly until it's done. On a VM with very poor performance this could go on for quite some time.

Reply:
I guess then the question is - is it "faster" for us to:

1. fail CI due to a slow disk (and then likely have a PR sit around for ~24h?) until a dev looks at logs, just to run `azp ci` again and gets a 10 minute run on a faster disk? Or
2. accept that we will occasionally get a long 60m `go generate` that eventually completes, or even never completes and has the CI stuck on the PR?

My guess is that even if we fail generate runs often, option 2 will be faster than option 1 for our FTS team, since most people reviewing PRs see :x: on CI and think "that's a YOU problem" not "that's a VM disk problem, I should re-run CI" :p

Alternatively, I'd be onboard with a timeout if we had automatic retry logic built in - because that will 100% of the time be the answer to getting unlucky like this.

Reply:
@SudoBrendan I like the idea of retry logic. I can update this PR to shift priority to retrying `go generate` upon failure, rather than failing the job run.  

Are you suggesting retry logic to retry go generate in the same job agent, or starting a new job run on a new agent?

If this is due to slow disk performance, is it possible to start the job on a new agent, from within the job run itself? I think the answer is yes, with github/ADO APIs.

Reply:
I would say if disk is the issue, we should spawn a new Job and/or retry the current Job, if that is possible.